tests/python_package_test/test_engine.py

```diff
@@ -2273,6 +2273,81 @@ def test_multiple_feval_train():
     assert 'decreasing_metric' in evals_result['valid_0']
 
 
+def test_objective_callable_train():
+    # Test classification
+    X, y = load_breast_cancer(return_X_y=True)
+    params = {'verbose': -1, 'objective': dummy_obj, 'metric': 'binary_logloss'}
+    X_train, X_validation, y_train, y_validation = train_test_split(X, y, test_size=0.2)
+    train_dataset = lgb.Dataset(data=X_train, label=y_train)
+    validation_dataset = lgb.Dataset(data=X_validation, label=y_validation, reference=train_dataset)
+    evals_result = {}
+    lgb.train(
+        params=params,
+        train_set=train_dataset,
+        valid_sets=validation_dataset,
+        num_boost_round=5,
+        feval=decreasing_metric,
+        callbacks=[lgb.record_evaluation(evals_result)]
+    )
+    assert len(evals_result['valid_0']) == 2
+    assert 'binary_logloss' in evals_result['valid_0']
+    assert 'decreasing_metric' in evals_result['valid_0']
```

Comment:
Per https://github.com/microsoft/LightGBM/pull/5052#issuecomment-1067492446, please remove tests related to custom metrics, and ensure that the tests added in this PR only contain the code necessary to test the following:

* if a callable is found in `params["objective"]` (or any of the aliases of `"objective"`), that objective is used
* if a callable is found in `params["objective"]` (or any of the aliases of `"objective"`) AND another one is passed via `fobj`, the objective provided to `params` is preferred to the one provided to `fobj`

Reply:
ok @jameslamb, I've added a test `test_choose_param_value_objective` to test this functionality within `_choose_param_value`. 

The reason for this is that `cv` returns a `Booster`, which has a `params` attribute, inside params `objective` should be `none`.

```python
cv_res = lgb.cv(
      params_with_metric,
      lgb_train,
      num_boost_round=10,
      nfold=3,
      stratified=False,
      return_cvbooster=True
)

cv_booster = cv_res['cvbooster'].boosters
cv_objs = [
    cb.params['objective'] == 'none' for cb in cv_booster
]
assert all(cv_objs)
```

Reply:
@jameslamb there's one caveat with `_choose_param_value`, when `objective` is string and `fobj` callable, the second one should take precedence.

Reply:
Can we remove `fjob` separate argument from `train()` and `cv()` functions right now in this PR given that our next `4.0.0` release will contain a lot of breaking changes?
https://github.com/microsoft/LightGBM/issues/3244#issuecomment-1056004279
@jameslamb @jmoralez @shiyu1994 @guolinke

Reply:
I support removing `fobj` right now, as part of this PR.

At this point, it still seems like v4.0.0 is a long way away and might contain many breaking changes. Given that it might be a long time until the 4.0.0 release, I don't think this project is able right now to do a deprecation warning followed by a removal in a separate release.

Reply:
Just a note: we should preserve this clarification of callable objective signature
https://github.com/microsoft/LightGBM/blob/d4cdbcfe1b1db2d48e95665a8b358028c496cdb2/python-package/lightgbm/engine.py#L54-L72

I think we can move to the bottom of the docstring and convert into a `Note` section similarly how it's done in the sklearn wrapper.

https://github.com/microsoft/LightGBM/blob/d4cdbcfe1b1db2d48e95665a8b358028c496cdb2/python-package/lightgbm/sklearn.py#L442

Reply:
Agree! @TremaMiguel can you please remove `fobj` from `train()` and `cv()`? You can update any tests that use it to instead take in a callable objective function via `params`.

Reply:
@jameslamb @StrikerRUS what is the ETA for `v4.0.0`? will this PR remain open until then?

Reply:
@TremaMiguel 
Unfortunately, there is no ETA yet. But because `v4.0.0` release will be the **next release** this PR will be merged into the `master` branch right after approving.

Reply:
got it thanks for the info